src/org/thoughtcrime/securesms/WebxdcActivity.java

```diff
@@ -192,6 +238,36 @@ private void updateTitleAndMenu() {
     });
   }
 
+  public static void addToHomeScreen(Activity activity, int msgId) {
+    Context context = activity.getApplicationContext();
+    try {
+      DcContext dcContext = DcHelper.getContext(context);
+      DcMsg msg = dcContext.getMsg(msgId);
+      final JSONObject info = msg.getWebxdcInfo();
+
+      final String docName = JsonUtils.optString(info, "document");
+      final String xdcName = JsonUtils.optString(info, "name");
+      byte[] blob = msg.getWebxdcBlob(JsonUtils.optString(info, "icon"));
+      ByteArrayInputStream is = new ByteArrayInputStream(blob);
+      BitmapDrawable drawable = (BitmapDrawable) Drawable.createFromStream(is, "icon");
+      Bitmap bitmap = drawable.getBitmap();
+
+      ShortcutInfoCompat shortcutInfoCompat = new ShortcutInfoCompat.Builder(context, "xdc-" + dcContext.getAccountId() + "-" + msgId)
+        .setShortLabel(docName.isEmpty() ? xdcName : docName)
+        .setIcon(IconCompat.createWithBitmap(bitmap))
+        .setIntents(getWebxdcIntentWithParentStack(context, msgId))
+        .build();
+
+      if (ShortcutManagerCompat.requestPinShortcut(context, shortcutInfoCompat, null)) {
+        Toast.makeText(context, R.string.done, Toast.LENGTH_LONG).show();
```

Comment:
Since we are just showing "Done", maybe use LENGTH_SHORT.

BTW, some apps go to the home screen after adding a shortcut so that the user can directly see it, but we don't have to do this.

Reply:
LENGTH_SHORT  makes sense, this is also what we are doing at the other places we feedback with a 'Done' toast. i added a commit.

> BTW, some apps go to the home screen after adding a shortcut so that the user can directly see it, but we don't have to do this.

which apps? going back to the homescreen is an idea worth trying out, not totally sure, however. but i would experiment with that maybe in another pr.

Reply:
thinking it over, i would not try to go the the homescreen, newer androids offer that option on their own:

<img width="329" alt="Screenshot 2022-08-15 at 11 25 50" src="https://user-images.githubusercontent.com/9800740/184611213-5e730946-0444-4a84-b438-32090425dd79.png">

here, the user can just drag the icon from the box and ends up in the homescreen, pretty nice ux (stock android 9).

if we mess around with finishing activities etc. we may easily introduce bugs and end up in handling corner cases and checking android versions ...

Reply:
> which apps? going back to the homescreen is an idea worth trying out, not totally sure, however.

Firefox, when you click "Add to Home Screen" in the three vertical dots menu.

In earlier times, the Play Store also did after installing an app, but it doesn't anymore. All in all, I think that I agree that we shouldn't do this, also because the Play Store apparently decided that they don' think it's good UX.